Re-review of PR #92: **not ready yet**. The shape is right, but I found two blockers in the current implementation.

## Blockers

1. **`default_host` can bypass the effective host policy when token `allowed_hosts` is unset**

`HostAccessManager.get_default_host()` returns `_request_default_host` when it is merely present in `_available_hosts` and, if a token scope exists, present in that scope:

```py
request_default = _request_default_host.get()
if request_default and request_default in self._available_hosts:
    if scope is None or request_default in scope:
        return request_default
```

That does not check the normal per-user/default host-access policy. For a token with `allowed_hosts = None`, `scope` is `None`, so any configured host can become the default even if the effective `HostAccessManager` policy would not allow that host. For a token with a scope and a matching real user entry, this also returns before intersecting with that user entry.

The request default should only win if it is in the **effective allowed hosts** for that request. Compute the effective allowed list first, then choose the request default only if it is in that list. Avoid calling `get_allowed_hosts()` recursively from inside `get_default_host()` unless you split out a helper.

Desired invariant:

```text
request default host may narrow/choose among allowed hosts; it must never expand host access
```

2. **Routes assume every identity has `.default_host`, breaking existing callers/tests**

The new code directly accesses `identity.default_host` in `/api/chat`, `/api/execute`, and WebSocket handling:

```py
token_default_host = identity.default_host if identity else ""
```

That breaks any injected/legacy identity object without the new attribute. Current targeted tests fail with 500s because the test identities are `SimpleNamespace` objects without `default_host`:

```text
FAILED tests/test_chat_session.py::test_chat_without_session_id_uses_identity_channel
FAILED tests/test_chat_session.py::test_chat_with_session_id_namespaces_channel
FAILED tests/test_chat_session.py::test_same_session_id_different_identities_do_not_collide
FAILED tests/test_chat_session.py::test_invalid_session_id_rejected_without_dispatch
FAILED tests/test_chat_session.py::test_permissions_come_from_identity_not_session

AttributeError: 'types.SimpleNamespace' object has no attribute 'default_host'
```

Use `getattr(identity, "default_host", "")` consistently. Odin has enough legacy identity paths already; no need to sacrifice another goat to AttributeError.

## Validation I ran

```text
python3 -m compileall -q src
python3 -m pytest tests/test_execute_api.py tests/test_chat_session.py tests/test_websocket_handler.py tests/test_web_chat.py -q
```

Result: compile passed; targeted pytest failed as above.

## Minor UX follow-up

If the user selects a default host, then changes host mode to `none` or unchecks that host in `select` mode, the stale `default_host` remains in the form and the API rejects on save. That is safe, but clunky. Clear `default_host` when it is no longer in the dropdown options.

Fix the policy bypass and the compatibility attr access, then rerun the targeted tests. The rest of the implementation matches the intended design.

--- a/src/permissions/host_access.py
+++ b/src/permissions/host_access.py
@@ -120,7 +120,8 @@ class HostAccessManager:
         scope = _request_host_scope.get()
         request_default = _request_default_host.get()
         if request_default and request_default in self._available_hosts:
-            if scope is None or request_default in scope:
+            effective = self.get_allowed_hosts(user_id)
+            if request_default in effective:
                 return request_default
         has_own_entry = user_id in self._users
         if scope is not None and not has_own_entry:

--- a/src/web/api.py
+++ b/src/web/api.py
@@ -894,7 +894,7 @@ def create_api_routes(bot: OdinBot) -> web.RouteTableDef:
         tier = identity.tier if identity else None
         token_tools = identity.allowed_tools if identity and identity.allowed_tools else None
         token_hosts = identity.allowed_hosts if identity and isinstance(getattr(identity, "allowed_hosts", None), list) else None
-        token_default_host = identity.default_host if identity else ""
+        token_default_host = getattr(identity, "default_host", "") if identity else ""
 
         # Optional caller-supplied session id for multi-request chat continuity.
         # Omitted -> historical behavior (one history per identity). Supplied -> validated
@@ -974,7 +974,7 @@ def create_api_routes(bot: OdinBot) -> web.RouteTableDef:
         token_tools = identity.allowed_tools if identity and identity.allowed_tools else None
         tier = identity.tier if identity else None
         token_hosts = identity.allowed_hosts if identity and isinstance(getattr(identity, "allowed_hosts", None), list) else None
-        token_default_host = identity.default_host if identity else ""
+        token_default_host = getattr(identity, "default_host", "") if identity else ""
 
         result = await process_web_chat(
             bot, content, channel_id,

--- a/src/web/websocket.py
+++ b/src/web/websocket.py
@@ -190,7 +190,7 @@ class WebSocketManager:
         tier = identity.tier if identity else None
         allowed_tools = identity.allowed_tools if identity and identity.allowed_tools else None
         token_hosts = identity.allowed_hosts if identity and isinstance(getattr(identity, "allowed_hosts", None), list) else None
-        token_default_host = identity.default_host if identity else ""
+        token_default_host = getattr(identity, "default_host", "") if identity else ""
 
         log.info("WebSocket chat from %s (tier=%s): %s", username, tier or "default", content[:80])
         try:

